File: mobile_client/app/src/main/java/wpam/mobile_client/protocol/PlotCommand.java
package wpam.mobile_client.protocol;

import java.util.ArrayList;

import wpam.mobile_client.sensor_tag.ParameterType;
import wpam.mobile_client.sensor_tag.SensorTagType;

public class PlotCommand extends Command
{
    private ParameterType parameterType;
    private SensorTagType sensorTagType;

    public PlotCommand(ParameterType parameterType, SensorTagType sensorTagType) {
        super(10, CommandType.PLOT);
        this.parameterType = parameterType;
        this.sensorTagType = sensorTagType;
    }
    public final void close()
    {
        super.close();
    }

    public ParameterType getParameterType() {
        return parameterType;
    }

    public void setParameterType(ParameterType parameterType) {
        this.parameterType = parameterType;
    }

    public SensorTagType getSensorTagType() {
        return sensorTagType;
    }

    public void setSensorTagType(SensorTagType sensorTagType) {
        this.sensorTagType = sensorTagType;
    }

    @Override
    public ArrayList<Integer> getFrameBytes()
    {
        initializeDataSize();

        ArrayList<Integer> frame = getHeader();
        frame.add(commandType_.getValue());
        frame.add(parameterType.getValue());
        frame.add(sensorTagType.getValue());

        return frame;
    }

    @Override
    public String getName()
    {
        return new String("SensorTagSamplesCommand");
    }

    @Override
    public void accept(CommandVisitor visitor)
    {
        visitor.visit(this);
    }

    @Override
    protected void initializeDataSize()
    {
        setDataSize(1);
    }
}

File: server/src/protocol/PlotCommand.cpp
#include "PlotCommand.h"
#include "CommandVisitor.h"

using namespace std;
using namespace communication;

PlotCommand::PlotCommand(const ParameterType parameterType, const SensorTagType sensorTagType)
        : Command(10, CommandType::PLOT),
          parameterType_(parameterType),
          sensorTagType_(sensorTagType)
{}

PlotCommand::~PlotCommand()
{}

vector<uint8_t> PlotCommand::getFrameBytes()
{
    initializeDataSize();

    vector<uint8_t > frame = getHeader();
    frame.push_back(static_cast<uint8_t>(commandType_));
    frame.push_back(static_cast<uint8_t>(parameterType_));
    frame.push_back(static_cast<uint8_t>(sensorTagType_));

    return frame;
}

std::string PlotCommand::getName()
{
    return string("PlotCommand");
}

void PlotCommand::accept(CommandVisitor &visitor)
{
    visitor.visit(*this);
}

void PlotCommand::initializeDataSize()
{
    uint16_t dataSize = sizeof(commandType_);
    dataSize += sizeof(parameterType_);
    dataSize += sizeof(sensorTagType_);

    setDataSize(dataSize);
}

ParameterType PlotCommand::getParameterType() const
{
    return parameterType_;
}

void PlotCommand::setParameterType(ParameterType parameterType)
{
    parameterType_ = parameterType;
}

SensorTagType PlotCommand::getSensorTagType() const
{
    return sensorTagType_;
}

void PlotCommand::setSensorTagType(SensorTagType sensorTagType)
{
    sensorTagType_ = sensorTagType;
}